In [1]:
# Cesarine Graham
# AST 5765 Homework 5
# 09/20/2025

In [2]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import os
import astropy.io.fits as fits 

#### 2. (10 Points) Here you will continue the sigma-clipping we worked on during Practicum3. Repeat the steps from problem 2b) on the subsample you created in 2b), to make a sub-subsample.

#### Print the new mean, median, and standard deviation. How different are the final mean and median?

#### How close is the final standard deviation to that expected for the Poisson distribution? Will this method always remove every bad pixel?

In [3]:
# number of photons
N = 10000

# creating initial array
data_ccd = np.zeros(400)
data_ccd[:396] = np.random.poisson(N, 396)
data_ccd[396:] = np.random.uniform(0, 1e6, 4)

print("Raw data mean:", np.mean(data_ccd))
print("Raw data median:", np.median(data_ccd))

# ----- first sigma clip (subsample) -----
sigma = np.std(data_ccd)
med_ccd = np.median(data_ccd)

low_range = med_ccd - 5*sigma
high_range = med_ccd + 5*sigma

filtered_data_ccd = data_ccd[(low_range < data_ccd) & (data_ccd < high_range)]

print("\nSubsample mean:", np.mean(filtered_data_ccd))
print("Subsample median:", np.median(filtered_data_ccd))
print("Subsample std dev:", np.std(filtered_data_ccd))

# ----- second sigma clip (sub-subsample) -----
sigma_2 = np.std(filtered_data_ccd)
med_2_ccd = np.median(filtered_data_ccd)

low_range_2 = med_2_ccd - 5*sigma_2
high_range_2 = med_2_ccd + 5*sigma_2

filtered_2_data_ccd = filtered_data_ccd[(low_range_2 < filtered_data_ccd) & (filtered_data_ccd < high_range_2)]

print("\nSub-subsample mean:", np.mean(filtered_2_data_ccd))
print("Sub-subsample median:", np.median(filtered_2_data_ccd))
print("Sub-subsample std dev:", np.std(filtered_2_data_ccd))

Raw data mean: 16437.65376169905
Raw data median: 10009.0

Subsample mean: 10271.54870546943
Subsample median: 10008.0
Subsample std dev: 5302.611585529911

Sub-subsample mean: 10005.123737373737
Sub-subsample median: 10007.5
Sub-subsample std dev: 93.46247743914314


#### The mean of the subsample (first sigma clip) and the sub-subsample (second sigma clip) cahnges from about 10272 to 10005. This is showing us that more outlier datapoints (bad-pixels) have been removed and the mean has shifted closer to the median, which is 10008. The medians have remained the same after each sigma clipping.

#### The final standard deviation is about 93.5. The standard deviation of a Poisson distribution is $\sqrt{10000} = 100$. Therefore, I think our final standard deviation is pretty close!

#### This method unfortunately will not remove every bad pixel. Some bad pixels will fall between the $5\sigma$ ranges.

#### In the Poisson Distribution, the standard deviation is equal to the square root of its mean. The standard deviation of our filtered dataset was calculated to be 3610.1, but the square of the mean is 100.9. The standard deviation of the unfiltered dataset was calculated to be roughly, 29579 which is VERY far off from the proper square root of the mean. 

#### Therefore, performing just masking the dataset with some sigma will not and has not removed every bad pixel.?